experimental/Schemes/Types.jl

```diff
@@ -985,3 +985,351 @@ mutable struct VarietyFunctionFieldElem{FracType<:AbstractAlgebra.Generic.Frac,
+        is_subset(U, domain(f)) || return false
+        is_subset(V, domain(g)) || return false
+        gU = preimage(g, U)
+        is_subset(gU, V) || return false
+      end
+      return true
+    end
+    function is_open_func(U::PrincipalOpenSubset, Y::AbsCoveredScheme)
+      return Y === X && ambient_scheme(U) in default_covering(X)
+    end
+    function is_open_func(U::AbsSpec, Y::AbsCoveredScheme)
+      return Y === X && U in default_covering(X)
+    end
+    function is_open_func(Z::AbsCoveredScheme, Y::AbsCoveredScheme)
+      return X === Y === Z
+    end
+    function is_open_func(U::AbsSpec, V::AbsSpec)
+      U in default_covering(X) || return false
+      V in default_covering(X) || return false
+      G = default_covering(X)[U, V]
+      return issubset(U, glueing_domains(G)[1])
+    end
+    function is_open_func(U::PrincipalOpenSubset, V::AbsSpec)
+      V in default_covering(X) || return false
+      ambient_scheme(U) === V && return true
+      W = ambient_scheme(U)
+      W in default_covering(X) || return false
+      G = default_covering(X)[W, V]
+      return is_subset(U, glueing_domains(G)[1])
+    end
+#    function is_open_func(U::PrincipalOpenSubset, V::PrincipalOpenSubset)
+#      ambient_scheme(V) in default_covering(X) || return false
+#      ambient_scheme(U) === ambient_scheme(V) && return issubset(U, V)
+#      W = ambient_scheme(U)
+#      W in default_covering(X) || return false
+#      G = default_covering(X)[W, ambient_scheme(V)]
+#      preV = preimage(glueing_morphisms(G)[1], V)
+#      return is_subset(U, preV)
+#    end
+    function is_open_func(W::SpecOpen, Y::AbsCoveredScheme)
+      return Y === X && ambient(W) in default_covering(X)
+    end
+    function is_open_func(W::SpecOpen, V::AbsSpec)
+      V in default_covering(X) || return false
+      ambient(W) === V && return true
+      U = ambient(W)
+      U in default_covering(X) || return false
+      G = default_covering(X)[U, V]
+      return is_subset(W, glueing_domains(G)[1])
+    end
+    function is_open_func(W::SpecOpen, V::PrincipalOpenSubset)
+      PW = ambient(W)
+      PV = ambient_scheme(V)
+      PW in default_covering(X) || return false
+      PV in default_covering(X) || return false
+      if PW === PV 
+        return issubset(W, V)
+        #return all(x->(issubset(x, V)), affine_patches(W))
+      else
+        G = default_covering(X)[PW, PV]
+        preV = preimage(glueing_morphisms(G)[1], V)
+        return issubset(W, preV)
+      end
+    end
+    function is_open_func(W::SpecOpen, V::SpecOpen)
+      PW = ambient(W)
+      PV = ambient(V)
+      PW in default_covering(X) || return false
+      PV in default_covering(X) || return false
+      if PW === PV 
+        return issubset(W, V)
+        #return all(x->(issubset(x, V)), affine_patches(W))
+      else
+        G = default_covering(X)[PW, PV]
+        preV = preimage(glueing_morphisms(G)[1], V)
+        return issubset(W, preV)
+      end
+    end
```

Comment:
For instance these two functions do exactly the same.
We can use julia's polymorphism here.

Reply:
That is true. And I'm surprised this is the case. One indicator that we did a reasonable job at streamlining the interfaces. 

But we can not hope for this to be true in general. For example, there is no good map type from a `SpecOpenRing` to another arbitrary ring. In this case, one has to wrap everything in a `MapFromFunc`. When we relate two `AbsSpec`s, on the other hand, we will want to build a homomorphism of rings to capture the relation. So I do not expect to find a lot of use cases for polymorphism in general.

Reply:
Okay. We can do some refactoring/cleanup later.